yjit/src/invariants.rs

```diff
@@ -257,24 +268,22 @@ pub extern "C" fn rb_yjit_root_mark() {
     }
 }
 
-/*
-static void
-yjit_block_assumptions_free(block_t *block)
-{
-    st_data_t as_st_data = (st_data_t)block;
-    if (blocks_assuming_stable_global_constant_state) {
-        st_delete(blocks_assuming_stable_global_constant_state, &as_st_data, NULL);
-    }
+/// Remove all invariant assumptions made by the block by removing the block as
+/// as a key in all of the relevant tables.
+pub fn block_assumptions_free(block: BlockRef) {
+    let invariants = Invariants::get_instance();
 
-    if (blocks_assuming_single_ractor_mode) {
-        st_delete(blocks_assuming_single_ractor_mode, &as_st_data, NULL);
+    for (block, bops) in invariants.block_basic_operators.iter() {
+        for (klass, bop) in bops.iter() {
+            invariants.basic_operator_blocks.get_mut(&(*klass, *bop)).map(|blocks| {
+                blocks.remove(&block);
+            });
+        }
```

Comment:
This is my lack of experience with rust showing. I thought `iter()` would run through the list and actually remove the elements from the list. If it doesn't, I can just add a `mem::take` here are that should fix it right?

Reply:
Also I don't get why there is a nested loop. I had assumed that the whole point of 

`block_basic_operators: HashMap<BlockRef, HashSet<(RedefinitionFlag, ruby_basic_operators)>>,`

is so you could do something like
```
let keys_for_blockref = block_basic_operators.get(blockref)?;
for key in &keys_for_blockref {
    basic_operator_blocks.get(key).map(|blockref_set| blockref_set.remove(blockref));
}
```

Reply:
It might be buggy to allow one block to assume the stability of multiple BOPs. If you do two BOPs A and B in sequence in a block, A can invalidate B and make us step into stale code since invalidation works at block boundaries. So to be more correct it should be:

`block_basic_operators: HashMap<BlockRef, (RedefinitionFlag, ruby_basic_operators)>`

at which point it might make more sense to store the tuple on the block rather than externally.

Reply:
How would a BOP invalidate another BOP? I'm not sure I understand that.